README.md
# video-engine

video engine react component integrated into dash for synced realtime playback of csv data and video

# NOTE

This repository is missing node_modules due to inability to upload so many files

# Note

In your pythong file import video-engine

Raw code for engine is located in/src

Any changes to /src should be saved and followed by:

npm run prepublish
python setup.py install

# Video

Your video needs to be hosted, It won't load locally into dash easiest way to host is:


npm install http-server
then run
http-server in the folder in which your video is hosted
I've uploaded a testvideo.mp4 which will work with CleanData.csv
# running dash

python test_vid.py will work

Your welcome to use my CSV data, its data pulled from a EEG, each entry is timestamp in relative time to ensure sync with video


# Problems

1. Can't get react to update video quicker then 500ms
2. multiple graphs become very slow and laggy
3. can't control video time with dash component, most likely needs to be fixed in the JS side
4. video/data seems to slowly become out of sync? should be tested more
5. CSS styling would be awesome.
6. I have to clean and format my EEG data before its usable, this is done in python in a second script
looking to add upload functionality for CSV, then have pandas procces the data, then plot it out on a graph, this cleaning procces also includes taking a realtime marker of where the video recording started thats planted into the CSV data and changing that to the first entry in the CSV data, also converting UNIX time to relative time, and applying filtering to the data.


# Why am I building this?

To allow myself to conduct my own studies using EEG/ECG's etc, I wanted a way to sync the data to whatever video I was watching or video of myself that was going on during the study.
I needed a method of playback to relate realtime events (video) to the data coming from my brain as those events where happening.
Looking to expand this application to playback all sorts of biometric data and be able to procces multiple videos (video of what I was looking at, video of myself, etc).
Ideally I'd like this to have ~.5ms of accuraccy 

# Contributing?
Thank you! I appreciate any help I can get :)


CHANGELOG.md
# Change Log for video-engine
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased

### Added
- Feature x

### Fixed
- Bug y

## 0.0.1 - 2018-03-07
- Initial release

[Unreleased]: https://github.com//video-engine/v0.0.1...HEAD


test_vid.py
import random

import dash
import datetime
from dash.dependencies import Input, Output, State, Event
import dash_core_components as dcc
import dash_html_components as html
import plotly.plotly as py
from plotly import graph_objs as go
from plotly.graph_objs import *
from flask import Flask
import pandas as pd
from collections import deque

import video_engine as rpd

app = dash.Dash()

app.scripts.config.serve_locally = True 
X = deque(maxlen=1)
Y = deque(maxlen=1)
X.append(1)
Y.append(1)

newtime = deque(maxlen=0)

df = pd.read_csv('CleanData.csv') #Reads CSV Data


app.layout = html.Div(children=[
    html.H1(children='the best fucking data analyzer ever',
            style={
                'textAlign': 'center'
                    }),
    dcc.Graph(id='sens1'),
 #   html.H2(id = 'time_counter',
     #    children='''NO TIME YET''',
   #      style={
  #              'textAlign': 'center'
   #                 }),
    html.Div(children=rpd.my_Player(
        id = 'video_player',
        url = 'http://127.0.0.1:8080/testvideo.mp4',
        width = 900,
        height = 720,
        controls = True,
        playing = False,
        seekTo = 0,
        volume = 1 ),
        style={'textAlign': 'center'}
    ),
dcc.Slider(id = 'time-slider', value=0, min=0, max=120, step=0.00001,
           marks={0: 'Start', 120: 'End'}),

    
    ])


@app.callback( ##Graph 1
    dash.dependencies.Output('sens1', 'figure'),
    [dash.dependencies.Input('video_player', 'currTime')])
def update_time(newtime):
    X.append(newtime)
    data = go.Scatter(
                x = df['timestamp'],
                y = df['sens1'],
                name = 'Scatter',
                mode = 'lines'
        )
    
    return {'data': [data], 'layout': go.Layout(xaxis = dict(range=[min(X), max(X)], fixedrange='true'),)}


# TESTING
#  Tested
#   Playing (True or False)
#   Volume (between 0 and 1)
#   seekTo
#   muted
#   playbackRate
#  To BE Tested
#   getCurrentTime
#   getDuration
#   controls (volume only)
#   styles
#   playsInline
#   config


app.css.append_css({
    'external_url': 'https://codepen.io/chriddyp/pen/bWLwgP.css'})

app.css.append_css({
    'external_url': 'https://unpkg.com/video-react@0.9.4/dist/video-react.css'})

if __name__ == '__main__':
    app.run_server(debug=True)


dash_test.py
import dash
from dash.dependencies import Input, Output, State
import dash_core_components as dcc
import dash_html_components as html
import pandas as pd
from flask import Flask
app = dash.Dash()
df = pd.read_csv('CleanData.csv')




if __name__ == '__main__':
    app.run_server(debug=True)

test_vid.1.py
import dash
from dash.dependencies import Output, Event, Input
import dash_core_components as dcc
import dash_html_components as html
from pandas_datareader.data import DataReader
import time
import plotly.graph_objs as go
from collections import deque
import plotly
from flask import Flask
import datetime
import pandas as pd
import random
import video_engine as rpd
# ^^ Dependences ^^ #

X = deque(maxlen=1)
Y = deque(maxlen=1)
X.append(1)
Y.append(1)

newtime = deque(maxlen=0)

app = dash.Dash("Beyond Data") #Creates application

df = pd.read_csv('CleanData.csv') #Reads CSV Data


app.layout = html.Div(children=[
    html.H1(children='Hello Dash'),

   dcc.Graph(id='data-graph'),
   html.H2(id='text', children='Nothing'),
    dcc.Slider(
        id='time-slider',
        value=0,
        min = 0,
        max = 120,
        step = 1,
        marks={0: 'Start', 120: 'End'}
    )
])


@app.callback( #Update graph
    dash.dependencies.Output('data-graph', 'figure'),
    [dash.dependencies.Input('time-slider', 'value')])
def update_time(newtime):
    X.append(newtime)
    data = go.Scatter(
                x = df['timestamp'],
                y = df['sens1'],
                name = 'Scatter',
                mode = 'lines'
        )
    
    return {'data': [data], 'layout': go.Layout(xaxis = dict(range=[min(X), max(X)], fixedrange='true'),)}

@app.callback( # Update Text
    dash.dependencies.Output('text', 'children'),
    [dash.dependencies.Input('time-slider', 'value')])
def update_text(newTime):
    return("Time - {}".format(str(newTime)))  

if __name__ == '__main__':
    app.run_server(debug=True)





external_css = ["https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/css/materialize.min.css"]
for css in external_css:
    app.css.append_css({"external_url": css})

external_js = ['https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/js/materialize.min.js']
for js in external_css:
    app.scripts.append_script({'external_url': js})


if __name__ == '__main__':
    app.run_server(debug=True)

src/components/my_Player.react.js
import React, { Component, PropTypes } from 'react';
// import { PrismCode } from 'react-prism';
import ReactPlayer from 'react-player';



// const sources = {
//   bunnyMovie: 'http://media.w3.org/2010/05/bunny/movie.mp4',
//     sailingMovie : '2017 Rolex Sydney Hobart Live Broadcast replay.mp4'
//     sailingMovie : 'file:///Users/ianburns/Documents/Code-Dev/PycharmProjects/react-video-component/2017 Rolex Sydney Hobart Live Broadcast replay.mp4'
// };

export default class my_Player extends Component {
  constructor(props, context) {
    super(props, context);
    // this.propsToState = this.propsToState.bind(this);
    // this.postUpdate = this.postUpdate.bind(this);
    this.getCurrentTime = this.getCurrentTime.bind(this);
    this.myTimer = this.myTimer.bind(this);
    this.myVar = setInterval(this.myTimer, 500);
  }

  //   //********
  //  propsToState(newProps) {
  //       /*
  //        * state includes:
  //        * - user modifiable attributes
  //        */
  //       // const {setProps, fireEvent} = this.props;
  //       // const {setProps} = this.props;
  //       // // const props = this.refs.player.props
  //       const newState = {};
  //       const currTime_now = this.refs.player.getCurrentTime();
  //       this.props['currTime'] = currTime_now
  //       // setProps({currTime: currTime_now});
  //       console.log('current Time in setProps = ' + String(currTime_now));
  //       this.postUpdate()
  //       // const my_Props = [
  //       //     'volume',
  //       // ]
  //       for(var prop in newProps) {
  //       // newProps.forEach(prop => {
  //           if (newProps[prop] !== undefined){
  //               switch (prop) {
  //                   // case 'muted':
  //                   //     console.log('in muter ');
  //                   //     this.props.muted = newProps[prop];
  //                   //     var x = newProps[prop];
  //                   //     console.log('in muted update = ' +  String(x) );
  //                   //     break;
  //                   // case 'volume':
  //                   //     console.log('BEFORE player volume = ' + String(this.refs.player.volume));
  //                   //     this.refs.player.volume = newProps['volume'].toFixed(1)
  //                   //     console.log(' AFTER player volume = ' + String(this.refs.player.volume));
  //                   //     break;
  //                   // case 'playbackRate':
  //                   //     console.log('BEFORE playbackRATE = ' + String(this.refs.player.playbackRate));
  //                   //     this.refs.player.playbackRate = newProps['playbackRate'].toFixed(1)
  //                   //     console.log(' AFTER playbackRate = ' + String(this.refs.player.playbackRate));
  //                   //     break;
  //                   case 'seekTo':
  //                       //console.log('BEFORE seek = ' + String(newProps['seekTo']));
  //                       this.refs.player.seekTo(newProps['seekTo']);
  //                       //console.log(' AFTER seek = ');
  //                       break;
  //                   // case prop:
  //                   //     console.log(' in case current Time = ' + String(currTime));
  //                   //     setProps({currentTime: currTime});
  //                   //     break;
  //                   default:
  //                        newState[prop] = newProps[prop];
  //               }
  //           }
  //       }
  //
  //  }
  // //
  // componentDidMount() {
  // //   // subscribe state change
  // //   //this.refs.player.subscribeToStateChange(this.handleStateChange.bind(this));
  //   const currTime_now = this.refs.player.getCurrentTime();
  //   console.log('currTime in did Mount ' + String(currTime_now))
  //   this.props['currTime'] = currTime_now
  //   // this.refs.player.setProps(currTime : 0)
  // }

    // componentWillReceiveProps(newProps) {
    //     // this.propsToState(newProps)
    //     // this.postUpdate(newProps)
    //     this.getCurrentTime(newProps)
    // }



    myTimer() {
    //  Console.log(' in myTimer ' + String(this.myVar) )
        this.getCurrentTime()
    }

    getCurrentTime(){
      // Console.log(' in postUpdate '  )
        // const {fireEvent} = this.props;
        const {setProps, fireEvent} = this.props;
        const currTime_now = this.refs.player.getCurrentTime();
      //  Console.log('currTime in postUpdate ' + String(currTime_now))
        if (setProps  !== null) {
      //      Console.log(' in in setProps ')
            setProps({currTime: currTime_now});
        }

        if (fireEvent) {
       //     Console.log('in FireEvent ' )
            fireEvent('change');
        }

    }

    // componentDidUpdate(prevProps, prevState){
    //    console.log(' in postUpdate ' + String(prevState) + String(prevProps) )
    //     const {setProps, fireEvent} = this.props;
    //     const currTime_now = this.refs.player.getCurrentTime();
    //     console.log('currTime in postUpdate ' + String(currTime_now))
    //     if (setProps  !== null) {
    //         setProps({currTime: currTime_now});
    //     }
    //
    //     if (fireEvent) {
    //          console.log('in change ' )
    //         fireEvent('change');
    //     }
    //
    // }


  render() {
      const{
          url,
          width,
          height,
          playing,
          playsInline,
          playbackRate,
          muted,
          volume
      } = this.props;


      // return <ReactPlayer url='https://www.youtube.com/watch?v=ysz5S6PUM-U' playing />
              // <url={this.state.source} />
        //       <pre>
        //   <PrismCode className="language-json">
        //     {JSON.stringify(this.state.player, null, 2)}
        //   </PrismCode>
        // </pre>


    return (

      <div>

        <ReactPlayer
          ref="player"

          getCurrentTime={this.getCurrentTime}
          width={width}
          height={height}
          url={url}
          playing={playing}
          playsInline={playsInline}
          playbackRate={playbackRate}
          muted={muted}
          volume={volume}

          controls = {true}
        >
        </ReactPlayer>

        <div>State</div>
              'URL = ' {url}
          '   Playing = '  {playing}
      </div>

    );
  }
}

my_Player.propTypes = {

	id: PropTypes.string,

    /**
     * width of video player
     */
      width: PropTypes.number,

    /**
     * height of video player
     */
      height: PropTypes.number,

    /**
     * src of video
     */
      url: PropTypes.string,

    /**
     * set true of false to start stop playing
     */
     playing: PropTypes.bool,


    /**
     * inline player
     */
      playsInline: PropTypes.bool,

       /**
     * SHow Controls?
     */
      controls: PropTypes.bool,


     /**
     * set playback rate - useful! multiple of normal = 1
     */
      playbackRate: PropTypes.number,

    /**
     * muted or not
     */
      muted: PropTypes.bool,

    /**
     * preset volume set number
     */
      volume: PropTypes.number,

    /**
     * seek time of video seconds
     */
      seekTo: PropTypes.number,

    /**
     *  curentTime from video
     */
      currTime: PropTypes.number,

    /**
     *  curentTime from video
     */
      getCurrentTime: PropTypes.number,


    /**
     * Dash-assigned callback that should be called whenever any of the
     * properties change
     */
    setProps: PropTypes.func,

    /**
     * Dash-assigned callback that gets fired when the value changes.
     */
    // dashEvents: PropTypes.oneOf(['click','change'])
    dashEvents: PropTypes.oneOf(['change'])


};




EEG_0.4.py
import random

import dash
import datetime
from dash.dependencies import Input, Output, State, Event
import dash_core_components as dcc
import dash_html_components as html
import plotly.plotly as py
from plotly import graph_objs as go
from plotly.graph_objs import *
from flask import Flask
import pandas as pd
from collections import deque

import video_engine as rpd
# ^^ Dependences ^^ #

X = deque(maxlen=20)
Y = deque(maxlen=20)
X.append(1)
Y.append(1)

app = dash.Dash("Beyond Data") #Creates application

df = pd.read_csv('testCleanREDO.csv') #Reads CSV Data


data_dict = {"data1":df['data1'],
"data2": df['data2'],
"data3": df['data3'],
"data4":df['data4']}

app.layout = html.Div([
    html.Div([
        html.H2('Beyond Data',
                style={'textAlign': 'center'
                       }),
        ]),
    dcc.Dropdown(id='eeg-data', #Creates dropdown menu for all options defined in dict
                 options=[{'label': s, 'value': s}
                          for s in data_dict.keys()],
                 value=[],
                 multi=True
                 ),
    dcc.Slider(id='time-slider', value=0, min=-10, max=20, step=0.00001,
                labels={-10: 'start', 20: 'end'}),
    rpd.my_Player(
        id = 'video_player',
        url = 'http://127.0.0.1:8080/testvideo.mp4',
        width = 900,
        height = 720,
        controls = True,
        playing = True ),             
    html.Div(children=html.Div(id='graphs'), className='row'), #Conatins all of the Graphs
    dcc.Interval( #Updates graphs determind in ms by interval var
        id='graph-update',
        interval=100),
    ], className="container",style={'width':'98%','margin-left':10,'margin-right':10,'max-width':50000})


@app.callback(
    dash.dependencies.Output('graphs','children'),
    [dash.dependencies.Input('eeg-data', 'value')],
    events=[dash.dependencies.Event('graph-update', 'interval')]
    )
def update_graph(data_names):
    graphs = []
    global X 
    global Y               
    X.append (X[-1]+1)
    Y.append (Y[-1]+1)
    if len(data_names)>2: 
        class_choice = 'col s12 m6 l4'
    elif len(data_names) == 2:
        class_choice = 'col s12 m6 l6'
    else:
        class_choice = 'col s12'


    for data_name in data_names:

        data = go.Scatter(
            x=list(df['timestamp']),
            y=list(data_dict[data_name]),
            name='Scatter',
            fill="tozeroy",
            fillcolor="#6897bb"
            )

        graphs.append(html.Div(dcc.Graph(
            id=data_name,
            animate=True,
            figure={'data': [data],'layout' : go.Layout(xaxis=dict(range=[min(X), max(X)]),
                                                        yaxis=dict(range=[min(data_dict[data_name]),max(data_dict[data_name])]),
                                                        margin={'l':50,'r':1,'t':45,'b':1}, #setting margins for Left,right,top,bototm
                                                        title='{}'.format(data_name))}
            ), className=class_choice))

    return graphs



external_css = ["https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/css/materialize.min.css"]
for css in external_css:
    app.css.append_css({"external_url": css})

external_js = ['https://cdnjs.cloudflare.com/ajax/libs/materialize/0.100.2/js/materialize.min.js']
for js in external_css:
    app.scripts.append_script({'external_url': js})


if __name__ == '__main__':
    app.run_server(debug=True)